File: 백준/Gold/13398. 연속합 2/연속합 2.py
import sys

#stdin = open("input.txt")

n = int(sys.stdin.readline().rstrip())
arr = list(map(int, sys.stdin.readline().split(" ")))

dp = [[0, 0] for _ in range(n)]

# 첫번째 인덱스: 중간 삭제 x
# 두번째 인덱스: 중간 삭제 o
dp[0] = [arr[0], arr[0]]
if n > 1:
    dp[1][0] = max(dp[0][0] + arr[1], arr[1])
    dp[1][1] = arr[1]

for i in range(2, n):
    dp[i][0] = max(dp[i-1][0] + arr[i], arr[i])
    dp[i][1] = max(dp[i-2][0], dp[i-1][1]) + arr[i]

maxValue = dp[0][0]
for i in range(n):
    if dp[i][0] > maxValue:
        maxValue = dp[i][0]
    if dp[i][1] > maxValue:
        maxValue = dp[i][1]

print(maxValue)

File: 백준/Gold/13398. 연속합 2/연속합 2.java
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int N = Integer.parseInt(br.readLine());
        String[] inputArr = br.readLine().split(" ");
        int[] arr = new int[N];
        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(inputArr[i]);
        }

        int[][] dp = new int[N][2];

        dp[0][0] = arr[0];
        dp[0][1] = arr[0];
        if (N > 1) {
            dp[1][0] = Math.max(dp[0][0] + arr[1], arr[1]);
            dp[1][1] = arr[1];
        }

        for (int i = 2; i < N; i++) {
            dp[i][0] = Math.max(arr[i], arr[i] + dp[i - 1][0]);
            dp[i][1] = Math.max(arr[i] + dp[i - 1][1], arr[i] + dp[i - 2][0]);
        }

        int maxValue = Integer.MIN_VALUE;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < 2; j++) {
                if (maxValue < dp[i][j]) {
                    maxValue = dp[i][j];
                }
            }
        }

        bw.write(maxValue + "\n");

        bw.flush();
    }
}